<a href="https://colab.research.google.com/github/kdrm12/AI-Bootcamp-Notebook/blob/main/05_database_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project: Database

**Instructions for Students:**

Please carefully follow these steps to complete and submit your project:

1. **Make a copy of the Project**: Please make a copy of this project either to your own Google Drive or download locally. Work on the copy of the project. The master project is **Read-Only**, meaning you can edit, but it will not be saved when you close the master project. To avoid total loss of your work, remember to make a copy.

2. **Completing the Project**: You are required to work on and complete all tasks in the provided project. Be disciplined and ensure that you thoroughly engage with each task.
   
3. **Creating a Google Drive Folder**: Each of you must create a new folder on your Google Drive. This will be the repository for all your completed project files, aiding you in keeping your work organized and accessible.
   
4. **Uploading Completed Project**: Upon completion of your project, make sure to upload all necessary files, involving codes, reports, and related documents into the created Google Drive folder. Save this link in the 'Student Identity' section and also provide it as the last parameter in the `submit` function that has been provided.
   
5. **Sharing Folder Link**: You're required to share the link to your project Google Drive folder. This is crucial for the submission and evaluation of your project.
   
6. **Setting Permission to Public**: Please make sure your Google Drive folder is set to public. This allows your instructor to access your solutions and assess your work correctly.

Adhering to these procedures will facilitate a smooth project evaluation process for you and the reviewers.

## Project Description

In this project, you'll work with the Kaggle E-commerce dataset. Your tasks are to import this dataset into a SQLite database, explore it using SQL to identify key attributes and patterns, and perform basic data cleaning procedures. This project is designed to improve your skills in database management, SQL usage, and data cleaning.

## Grading Criteria

There are 13 criterias in this project, all have the same weight divided into two: 8 criterias for SQL and 5 criterias for Elasticssearch. Each task will give you either 100 point if you are correct and 0 if you are wrong. The final score for the project will the the average of all 13 tasks.

* Criteria 1, Task I.3: This task will assess your ability to create an SQL database, in this case using sqlite. To ensure correct data is imported, this will check that database table have the correct amount of columns and the correct column names. This will ensure a uniform dataset for all students.

* Criteria 2, Task I.3: Same as Criteria 1, instead of columns, this will check that the total rows of data in the table is correct. This will ensure a uniform dataset for all students.

* Criteria 3, Task I.3.4 query 1: The task will assess your ability to understand and perform basic dataset check, that one row in the database does not guarantee uniqueness for individual columns.

* Criteria 4, Task I.3.4 query 1: Same as Criteria 3.

* Criteria 5, Task I.3.4 query 2: This task will assess your ability to create a SQL query that will get the data, perform arithmetic operations on the data, sort and limit the data returned.

* Criteria 6, Task I.3.4 query 3: Same as Criteria 5.

* Criteria 7, Task I.3.4 query 4: This task will assess your ability to create a SQL query that will get the data, perform arithmetic operations on the data, sort, limit the data and structure the result in a concise format where the all the data are retrieved and presented in a format that is easy to understand can be further processed through programming.

* Criteria 8, Task I.3.4 query 5: Same as Criteria 7.

* Criteria 9, Task II.3: This task will assess your ability to create an Elasticsearch Index. To ensure correct data is imported, this will check that the index have the correct amount of columns the the correct column names. This will ensure a uniform dataset for all students.

* Criteria 10, Task II.3: Same as Criteria 9, instead of columns, this will check that the total rows of data in the table is correct. This will ensure a uniform dataset for all students.

* Criteria 11, Task II.4 query 1: This task will assess your ability to perform a basic elastic search and return the result.

* Criteria 12, Task II.4 query 2: This task will assess your ability to perform an elastic search, perform arithmetic operations on the data and return the result.

* Criteria 13, Task II.4 query 3: This task will assess your ability to perform an elastic search, perform arithmetic operations on the data, sort and limit the data returned.


## Student Identity

In [ ]:
# @title #### Student Identity
student_id = "REAABPRB" # @param {type:"string"}
name = "Kevin Adrian Manurung" # @param {type:"string"}
drive_link = "https://drive.google.com/drive/folders/1dg14HN9pEW8shaCH7sHyW2tCB7DP1Tnq?usp=sharing"  # @param {type:"string"}

assignment_id = "00_database_project"

# Import grader package
!pip install rggrader
from rggrader import submit, submit_image

Rmember to make a copy of this notebook in your Google Drive and work in your own copy.

## I. Loading and Querying with SQL

In this task, you will use Kaggle E-commerce dataset. You need to download it from the link here: [Kaggle E-commerce data](https://www.kaggle.com/datasets/carrie1/ecommerce-data?select=data.csv).

After you download the dataset, you will then import the dataset into SQL format, using SQLite.

### I.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [ ]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below

import pandas as pd
import sqlite3

### I.2. Data Loading and SQL CREATE and INSERT to database table

Now you can create the database and tables. Follow the same column naming as the dataset.

Next insert all the data from the dataset into the table in the database.

In [ ]:
# Write your data loading to SQL database here

# read data to dataframe
df = pd.read_csv('https://drive.google.com/uc?id=169I8LMReQIUQoGGELcDW3hITOP6YE2gm', encoding='latin1')

# make database .db
conn = sqlite3.connect('ecommerce.sqlite')
df.to_sql("data", conn, if_exists="replace", index=False)
conn.commit()
conn.close()


### I.3. Verifying the database

Before we go to the next section, let's check our database first. You need to change the db_name and table_name with your database name and table name, afterwards you can just run the code block. The result should be:

- Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
- Number of rows: 541909

In [ ]:
import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('ecommerce.sqlite') #change db_name to your database name.
cur = conn.cursor()

# Get column names
cur.execute("PRAGMA table_info(data)") #change table_name to your table name
sql_columns = [column[1] for column in cur.fetchall()] #the variable sql_columns is used for submission, do not change the variable name
print(f"Columns: {sql_columns}")

# Get total number of rows
cur.execute("SELECT COUNT(*) FROM data") #change table_name to your table name
sql_num_rows = cur.fetchone()[0] #the variable sql_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {sql_num_rows}")

# Close the connection
conn.close()

Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
Number of rows: 541909


### I.3.4. Running SQL Queries

Once we have the database in SQL format, let's do some data exploration and analysis.

You are part of the engineering team at a global e-commerce company. Your company has a vast catalogue of products and serves customers around the globe. As the company continues to grow, the management decides it's time to expand the business even more.

As an essential part of this expansion, your team is tasked with diving into your company's rich e-commerce dataset to extract crucial insights that will guide the expansion strategy.

1. **Total number of unique products and unique customers:**

   Your first task is to gauge the breadth of your operations. By determining the total number of unique products you sell and the total number of unique customers you serve, you can assess the scale and diversity of your business.

2. **Total revenue for each product:**

   Next, you're going to identify the star performers in your product portfolio. Calculating the total revenue for each product will reveal which items are the biggest revenue drivers. These insights can help guide decisions about product focus and marketing efforts. Limit the result to the top 5 product with the biggest revenue.

3. **Most profitable countries (by total sales):**

   To pinpoint the most fruitful geographic areas for your business, you'll identify which countries generate the most profit. This data will help you understand where your operations are already strong and where there might be potential for regional expansion. Limit the result to the top 5 countries with the biggest profit.

4. **Top-selling products (by total sales) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by sales for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

4. **Top-selling products (by quantity) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by quantity for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

By answering these queries, your team will provide valuable data-driven insights that will directly contribute to the strategic decisions about where and how your business expands. As part of the engineering team, you're not just maintaining the technical infrastructure - you're shaping the future of the company.

> Hint: The Product name is stored in column Descrioption

In [ ]:
# Write your SQL Query here

#use the following variable name to store the result from your SQL Query
sql_num_products = 0
sql_num_customers = 0
sql_top_product_revenues = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_countries = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_selling_products_by_country_by_sales = {} #dictionary where key is country name, value is a list of top-selling products by sales, see example below
sql_top_selling_products_by_country_by_quantity = {} #dictionary where key is country name, value is a list of top-selling products by quantity, see example below


# 1. Find the total number of unique products and unique customers
# Example: sql_num_products = 1559
# Example: sql_num_customers = 1675
conn = sqlite3.connect('ecommerce.sqlite')
sql_num_products = pd.read_sql_query("SELECT COUNT(DISTINCT StockCode) AS NumProduct FROM data", conn)['NumProduct'].iloc[0]
print(f'sql_num_products = {sql_num_products}')
sql_num_customers = pd.read_sql_query("SELECT COUNT(DISTINCT CustomerID) AS NumCustomer FROM data", conn)['NumCustomer'].iloc[0]
print(f'sql_num_customers = {sql_num_customers}')

sql_num_products = 4070
sql_num_customers = 4372


In [ ]:
# 2. Calculate the total revenue for each product, limited to top 5
# Example: sql_top_product_revenues = [('PRODUCT 1', 149385.12), ('PRODUCT 2', 54895.53), ('PRODUCT 3', 44545.55), ('PRODUCT 4', 38000.00), ('PRODUCT 5', 36000.00)]
top_products_revenues_df = pd.read_sql_query("SELECT Description, SUM(Quantity*UnitPrice) AS TotalRevenue from data GROUP BY Description ORDER BY TotalRevenue DESC LIMIT 5", conn)
sql_top_product_revenues = list(zip(top_products_revenues_df['Description'], top_products_revenues_df['TotalRevenue']))
print(f'sql_top_product_revenues = {sql_top_product_revenues}')


sql_top_product_revenues = [('DOTCOM POSTAGE', 206245.48000000019), ('REGENCY CAKESTAND 3 TIER', 164762.19), ('WHITE HANGING HEART T-LIGHT HOLDER', 99668.46999999849), ('PARTY BUNTING', 98302.98000000048), ('JUMBO BAG RED RETROSPOT', 92356.03000000204)]


In [ ]:
# 3. Identify the most profitable countries (by total sales), limited to top 5
# Example: sql_top_countries = [('COUNTRY 1', 600000.00), ('COUNTRY 2', 334857.45), ('COUNTRY 3', 245879.00), ('COUNTRY 4', 180000.00), ('COUNTRY 5', 164389.45)]
top_countries_df = pd.read_sql_query("SELECT Country, SUM(Quantity*UnitPrice) AS Profit from data GROUP BY Country ORDER BY Profit DESC LIMIT 5", conn)
sql_top_countries = list(zip(top_countries_df['Country'], top_countries_df['Profit']))
print(f'sql_top_countries = {sql_top_countries}')


sql_top_countries = [('United Kingdom', 8187806.364001113), ('Netherlands', 284661.54000000015), ('EIRE', 263276.81999999826), ('Germany', 221698.20999999862), ('France', 197403.90000000002)]


In [ ]:
# 4. Find the top-selling products (by total sales) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_sales = {
#    'COUNTRY 1': [('PRODUCT 1', 200000.00), ('PRODUCT 2', 100000.00), ('PRODUCT 3', 95348.00), ('PRODUCT 4', 90000.43), ('PRODUCT 5', 80000.00)],
#    'COUNTRY 2': [('PRODUCT 1', 9000.00), ('PRODUCT 2', 7345.00), ('PRODUCT 3', 6934.23), ('PRODUCT 4', 6800.23), ('PRODUCT 5', 4000.00)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('PRODUCT 1', 7442.84), ('PRODUCT 2', 5175.00), ('PRODUCT 3', 3098.0), ('PRODUCT 4', 3014.34), ('PRODUCT 5', 2857.83)],
for country, _ in sql_top_countries:
  top_selling_products_by_country_by_sales_df = pd.read_sql_query(f"SELECT Description, SUM(Quantity*UnitPrice) AS TotalSales from data WHERE Country = '{country}' GROUP BY Description ORDER BY TotalSales DESC LIMIT 5", conn)
  sql_top_selling_products_by_country_by_sales[country] = list(zip(top_selling_products_by_country_by_sales_df['Description'], top_selling_products_by_country_by_sales_df['TotalSales']))
print(f'sql_top_selling_products_by_country_by_sales = {sql_top_selling_products_by_country_by_sales}')


sql_top_selling_products_by_country_by_sales = {'United Kingdom': [('DOTCOM POSTAGE', 206245.48000000019), ('REGENCY CAKESTAND 3 TIER', 134405.93999999992), ('WHITE HANGING HEART T-LIGHT HOLDER', 93953.0699999987), ('PARTY BUNTING', 92501.73000000033), ('JUMBO BAG RED RETROSPOT', 84516.44000000166)], 'Netherlands': [('RABBIT NIGHT LIGHT', 9568.48), ('ROUND SNACK BOXES SET OF4 WOODLAND ', 7991.400000000001), ('SPACEBOY LUNCH BOX ', 7485.5999999999985), ('DOLLY GIRL LUNCH BOX', 6828.599999999999), ('ROUND SNACK BOXES SET OF 4 FRUITS ', 4039.2000000000007)], 'EIRE': [('REGENCY CAKESTAND 3 TIER', 7442.849999999997), ('CARRIAGE', 5175.0), ('JAM MAKING SET WITH JARS', 3089.0), ('3 TIER CAKE TIN RED AND CREAM', 3041.5499999999997), ('WHITE HANGING HEART T-LIGHT HOLDER', 2857.8)], 'Germany': [('POSTAGE', 20821.0), ('REGENCY CAKESTAND 3 TIER', 8257.349999999997), ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3554.6999999999985), ('ROUND SNACK BOXES SET OF 4 FRUITS ', 1949.9500000000016), ('SPACEBOY L

In [ ]:
# 5. Find the top-selling products (by quantity) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_quantity = {
#    'COUNTRY 1': [('COUNTRY 1, PRODUCT 1', 22343), ('COUNTRY 1, PRODUCT 2', 12345), ('COUNTRY 1, PRODUCT 3', 9534), ('COUNTRY 1, PRODUCT 4', 9000), ('COUNTRY 1, PRODUCT 5', 8000)],
#    'COUNTRY 2': [('COUNTRY 2, PRODUCT 1', 9000), ('COUNTRY 2, PRODUCT 2', 7345), ('COUNTRY 2, PRODUCT 3', 6934), ('COUNTRY 2, PRODUCT 4', 6800), ('COUNTRY 2, PRODUCT 5', 4000)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5, PRODUCT 1', 48345), ('COUNTRY 5, PRODUCT 2', 43965), ('COUNTRY 5, PRODUCT 3', 43556), ('COUNTRY 5, PRODUCT 4', 34567), ('COUNTRY 5, PRODUCT 5', 33436)],
for country, _ in sql_top_countries:
  top_selling_products_by_country_by_quantity_df = pd.read_sql_query(f"SELECT Country, Description, SUM(Quantity) AS Quantity from data WHERE Country = '{country}' GROUP BY Description ORDER BY Quantity DESC LIMIT 5", conn)
  sql_top_selling_products_by_country_by_quantity[country] = list(zip(top_selling_products_by_country_by_quantity_df['Country'], top_selling_products_by_country_by_quantity_df['Description'], top_selling_products_by_country_by_quantity_df['Quantity']))
print(f'sql_top_selling_products_by_country_by_quantity = {sql_top_selling_products_by_country_by_quantity}')



sql_top_selling_products_by_country_by_quantity = {'United Kingdom': [('United Kingdom', 'WORLD WAR 2 GLIDERS ASSTD DESIGNS', 48326), ('United Kingdom', 'JUMBO BAG RED RETROSPOT', 43167), ('United Kingdom', 'POPCORN HOLDER', 34365), ('United Kingdom', 'ASSORTED COLOUR BIRD ORNAMENT', 33679), ('United Kingdom', 'WHITE HANGING HEART T-LIGHT HOLDER', 33193)], 'Netherlands': [('Netherlands', 'RABBIT NIGHT LIGHT', 4801), ('Netherlands', 'SPACEBOY LUNCH BOX ', 4528), ('Netherlands', 'DOLLY GIRL LUNCH BOX', 4132), ('Netherlands', 'PACK OF 72 RETROSPOT CAKE CASES', 4128), ('Netherlands', 'ROUND SNACK BOXES SET OF4 WOODLAND ', 3132)], 'EIRE': [('EIRE', 'PACK OF 72 RETROSPOT CAKE CASES', 1728), ('EIRE', '60 TEATIME FAIRY CAKE CASES', 1536), ('EIRE', 'VINTAGE SNAP CARDS', 1492), ('EIRE', 'ASSORTED INCENSE PACK', 1440), ('EIRE', 'PACK OF 60 PINK PAISLEY CAKE CASES', 1296)], 'Germany': [('Germany', 'ROUND SNACK BOXES SET OF4 WOODLAND ', 1218), ('Germany', 'ASSORTED COLOURS SILK FAN', 1164), ('Germa

## II. Indexing and Querying with Elasticsearch

You should have a better understanding of how to use SQL, now let's leverage the power of Elasticsearch to explore and analyze the data. We'll use the same dataset.


### II.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [ ]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below
%%bash

wget -q https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.9.2-linux-x86_64.tar.gz
wget -q https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.9.2-linux-x86_64.tar.gz.sha512
tar -xzf elasticsearch-oss-7.9.2-linux-x86_64.tar.gz
sudo chown -R daemon:daemon elasticsearch-7.9.2/
shasum -a 512 -c elasticsearch-oss-7.9.2-linux-x86_64.tar.gz.sha512


elasticsearch-oss-7.9.2-linux-x86_64.tar.gz: OK


In [ ]:
%%bash --bg

sudo -H -u daemon elasticsearch-7.9.2/bin/elasticsearch

In [ ]:
import time
time.sleep(20)

In [ ]:
%%bash

ps -ef | grep elasticsearch

root        1236    1234  0 03:34 ?        00:00:00 sudo -H -u daemon elasticsearch-7.9.2/bin/elasti
daemon      1237    1236 99 03:34 ?        00:00:21 /content/elasticsearch-7.9.2/jdk/bin/java -Xshar
root        1540    1538  0 03:35 ?        00:00:00 grep elasticsearch


In [ ]:
%%bash

curl -sX GET "localhost:9200/"

{
  "name" : "cc3f295c8717",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "2h28NVRKTXez1Av3IyBJCw",
  "version" : {
    "number" : "7.9.2",
    "build_flavor" : "oss",
    "build_type" : "tar",
    "build_hash" : "d34da0ea4a966c4e49417f2da2f244e3e97b4e6e",
    "build_date" : "2020-09-23T00:45:33.626720Z",
    "build_snapshot" : false,
    "lucene_version" : "8.6.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}


In [ ]:
%pip install elasticsearch==7.9.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 213.5/213.5 kB 4.0 MB/s eta 0:00:00


In [ ]:
from elasticsearch import Elasticsearch, helpers
import pandas as pd
from collections import Counter

### II.2. Set up Elasticsearch Index

The first step you need to do which is **preprocessing, let's keep it simple and simply drop all row that have missing values.**

Next, you need to set up an Elasticsearch index. An Elasticsearch index is like a database in a traditional relational database. An index lets you store, search, and analyze big volumes of data quickly and in near real time.

You'll also need to setup the mappings for the data types, let's do that here.

In [ ]:
# Preprocessing here

# Load your dataset and create an Elasticsearch index
# Hints: use the BulkAPI for faster creation of the index
df = pd.read_csv('https://drive.google.com/uc?id=169I8LMReQIUQoGGELcDW3hITOP6YE2gm', encoding='latin1')
df = df.dropna()

# create connection to elasticsearch
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

# create index
es.indices.create(index='ecommerce_data', body = {
    "mappings": {
        "properties": {
            "InvoiceNo": {
                "type": "keyword"
            },
            "StockCode": {
                "type": "keyword"
            },
            "Description": {
                "type": "text"
            },
            "Quantity": {
                "type": "integer"
            },
            "InvoiceDate": {
                "type": "date",
                "format": "M/d/yyyy H:m"
            },
            "UnitPrice": {
                "type": "float"
            },
            "CustomerID": {
                "type": "keyword"
            },
            "Country": {
                "type": "keyword"
            }
        }
    }
})

# convert data to dict
dict_data = df.to_dict('records')

# create bulk action
bulk_action =  [
    {
    "_index": 'ecommerce_data',
    "_id": i, # Use a unique identifier for the document
    "_source": record, # Document content
    }
    for i, record in enumerate(dict_data)]

try:
    response = helpers.bulk(es, bulk_action)
    print("Bulk upload completed:", response)
except helpers.BulkIndexError as e:
    print("Error during bulk upload:", e)

Bulk upload completed: (406829, [])


### II.3. Verifying the index

Before we go to the next section, let's check our index first. You need to change the my_index_name with your chosen index name, afterwards you can just run the code block. The result should be:

- Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
- Number of rows: 406829

In [ ]:
from elasticsearch import Elasticsearch

# Connect to the Elasticsearch server
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

index_name = "ecommerce_data" #change the my_index_name value

# Get index mapping (equivalent to getting column names in SQL)
mapping = es.indices.get_mapping(index=index_name)
es_columns = list(mapping[index_name]['mappings']['properties'].keys()) #the variable es_columns is used for submission, do not change the variable name
print(f"Columns: {es_columns}")

# Count documents in the index (equivalent to counting rows in SQL)
es_num_rows = es.count(index=index_name)['count'] #the variable es_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {es_num_rows}")


Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
Number of rows: 406829


### II.4. Running Elasticsearch queries

Once your e-commerce data is cleaned and ready, let's analyse the data to gain insights that will drive the expansion strategy, this time using Elasticsearch and we'll focus on one specific country, which is Germany.

1. **Find all transactions for a specific country**

   Your company operates globally, but you want to understand better how different countries contribute to your sales. By finding all transactions that happened in a specific country, you can get a clearer picture of your company's reach and performance in that location. In our case, we'll focus on the country Germany.


2. **Find all unique products in Germany**

   Your first task is to understand the breadth of your company's product portfolio in Germany. By determining the total number of unique products you sell, you can assess the scale and diversity of your operations.


3. **Find the top 5 most purchased products in Germany**

   Next up, you'll identify the best-selling products in your portfolio in Germany. Knowing which items are the most purchased can help guide decisions about product focus and marketing efforts. We'll limit the result to 5 most purchased products.


By executing these tasks, you will enable your team to provide data-driven insights which will directly contribute to the strategic decisions about your business expansion. As part of the data engineering team, you're not only maintaining the technical infrastructure - you're shaping the future of the company.


In [ ]:
# Write your Elasticsearch Query here

#use the following variable name to store the result from your Elasticsearch Query
es_total_unique_products = 0
es_transactions_in_germany = 0
es_top_products = {} #dictionary where key is product name, value is the quantity sold, see example below

# 1. Find all transactions for a specific country (Germany)
# Example: es_transactions_in_germany = 7584
response = es.search(index=index_name, body={
    "query": {
        "match": {
            "Country": "Germany"
        }
    }
})

es_transactions_in_germany = response['hits']['total']['value']
print(f"es_transactions_in_germany = {es_transactions_in_germany}")



es_transactions_in_germany = 9495


In [ ]:
# 2. Find all unique products in Germany
# Hints: Use set() to remove duplicate values
# Example: es_total_unique_products = 45345

response = es.search(index=index_name, body={
    "size": 10000,
    "query": {
        "match": {
            "Country": "Germany"
        }
    }
})
all_data = [hit['_source']['Description'] for hit in response['hits']['hits']]
products = set(all_data)
es_total_unique_products = len(products)
print(f"es_total_unique_products = {es_total_unique_products}")


es_total_unique_products = 1703


In [ ]:
# 3. Find the top 5 most purchased products in Germany
# Example: es_top_products = {'PRODUCT 1': 498, 'PRODUCT 2': 452, 'PRODUCT 3': 342, 'PRODUCT 4': 231, 'PRODUCT 5': 123}
count = Counter(all_data)
es_top_products = dict(count.most_common(5))

print(f'es_top_products = {es_top_products}')

es_top_products = {'POSTAGE': 383, 'ROUND SNACK BOXES SET OF4 WOODLAND ': 120, 'REGENCY CAKESTAND 3 TIER': 81, 'ROUND SNACK BOXES SET OF 4 FRUITS ': 78, 'PLASTERS IN TIN WOODLAND ANIMALS': 67}


## Submission

Once you are satisfied with the performance of your model, then you run the code block below to submit your project.


In [ ]:
# Submit Method

# Do not change the code below
question_id = "01_sql_columns"
submit(student_id, name, assignment_id, str(sql_columns), question_id, drive_link)
question_id = "02_sql_num_rows"
submit(student_id, name, assignment_id, str(sql_num_rows), question_id, drive_link)
question_id = "03_sql_num_products"
submit(student_id, name, assignment_id, str(sql_num_products), question_id, drive_link)
question_id = "04_sql_num_customers"
submit(student_id, name, assignment_id, str(sql_num_customers), question_id, drive_link)
question_id = "05_sql_top_product_revenues"
submit(student_id, name, assignment_id, str(sql_top_product_revenues), question_id, drive_link)
question_id = "06_sql_top_countries"
submit(student_id, name, assignment_id, str(sql_top_countries), question_id, drive_link)
question_id = "07_sql_top_selling_products_by_country_by_sales"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_sales), question_id, drive_link)
question_id = "08_sql_top_selling_products_by_country_by_quantity"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_quantity), question_id, drive_link)

question_id = "09_es_columns"
submit(student_id, name, assignment_id, str(es_columns), question_id, drive_link)
question_id = "10_es_num_rows"
submit(student_id, name, assignment_id, str(es_num_rows), question_id, drive_link)
question_id = "11_es_total_unique_products"
submit(student_id, name, assignment_id, str(es_total_unique_products), question_id, drive_link)
question_id = "12_es_transactions_in_germany"
submit(student_id, name, assignment_id, str(es_transactions_in_germany), question_id, drive_link)
question_id = "13_es_top_products"
submit(student_id, name, assignment_id, str(es_top_products), question_id, drive_link)

'Assignment successfully submitted'

## FIN